lib/segment/src/id_tracker/simple_id_tracker.rs

```diff
@@ -343,6 +345,28 @@ impl IdTracker for SimpleIdTracker {
         }
     }
 
+    fn iter_random(&self) -> Box<dyn Iterator<Item = (PointIdType, PointOffsetType)> + '_> {
+        let rng = rand::thread_rng();
+        let available = self.internal_to_external.len();
+        let uniform = rand::distributions::Uniform::new(0, available);
+        let iter = Distribution::sample_iter(uniform, rng)
+            // TODO: this is not efficient if `available` is large and we iterate over most of them,
+            // but it's good enough for low limits.
+            //
+            // We could improve it by using a variable-period PRNG to adjust depending on the number of available points.
+            .unique()
+            .take(available)
```

Comment:
A simple alternative may be, for each iteration:
- generate a random number
- if we haven't seen it yet, use it
- if we did see it, pick the next unseen number

It'll be faster, at the cost of being a bit less random.

Reply:
that's exactly what this is doing. The thing I am pointing out is that, let's say, the segment has 10k points, and you request a limit of 10k too. For finding the last number, it would have a 1/10k probability of having a random number which we haven't seen before. 

Ideally the best algorithm would be one which is pseudo-random, and doesn't repeat values until all of them have appeared once.

Reply:
> that's exactly what this is doing

It does not. I meant something different.

Your implementation keeps generating random numbers until we have a lucky unseen number.

I meant looping through the seen list in-order to find the next (ordered) unseen number without any randomness.

More like this, (but proper, and wrapping):
```rust
let num = random();
if is_already_seen {
    return (num + 1..).filter(|n| !seen.contains(n)).next().unwrap();
}
```

Reply:
Ah, I see what you mean now. However this would tend to have runs of sequential outputs, doesn't it?

Reply:
> However this would tend to have runs of sequential outputs, doesn't it?

Yes.

We might need some heuristic here to dynamically select the right approach. Because the 10k from 10k example you give would be extremely expensive.

@generall may have a good suggestion here.